### PR TITLE
[pt1][quant] Change fbgemm_linear_{int8,fp16}_weight to fbgemm_linear_{int8,fp16}_weight_fp32_activation

### PR DESCRIPTION
**Pull request by jianyuh**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#22955 [pt1][quant] Change fbgemm_linear_{int8,fp16}_weight to fbgemm_linear_{int8,fp16}_weight_fp32_activation**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D16297512/)

Following the comment in https://github.com/pytorch/pytorch/pull/22891, change the fbgemm wrapper function name to indicate whether it is dynamic quantization or static quantization.

Differential Revision: [D16297512](https://our.internmc.facebook.com/intern/diff/D16297512/)

### CODE REVIEW

**Review by jamesr66a — CHANGES_REQUESTED** [review]
This is a backward-incompatible change that is going to break existing models

File: aten/src/ATen/native/native_functions.yaml
Comment (jamesr66a) [inline aten/src/ATen/native/native_functions.yaml]: To preserve backward-compatibility, I suggest the following:

1) Keep `fbgemm_linear_{int8,fp16}_weight` functions.
2) Create separate `fbgemm_linear_{int8,fp16}_weight_fp32_activation` schema and functions.
3) Move implementations from the original functions into the `_fp32_activation` functions.
4) Have the original functions call directly into the `_fp32_activation` functions, and add a TORCH_WARN() call warning that those functions are deprecated

Comment (jamesr66a) [reply aten/src/ATen/native/native_functions.yaml]: Once we can ensure that all internal users have migrated off the old function names, we can then delete the old function schemas and implementations

Comment (jianyuh) [reply aten/src/ATen/native/native_functions.yaml]: Thanks @jamesr66a ! Updated the code with 1, 2, 3, 4 according to your suggestions.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jspark1105) [inline aten/src/ATen/native/QuantizedLinear.cpp]: From here, the implementation is same as fbgemm_linear_int8_weight_fp32_activation . Maybe we can just call that function after printing deprecation warning? (I think this is what James suggested)

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jspark1105) [inline aten/src/ATen/native/QuantizedLinear.cpp]: ditto

**Review by dzhulgakov — COMMENTED** [review]
honestly, I'm not sure it's worth complexity and wrappers. We could just keep the original names, couldn't we? You can just add comment in native_function.yaml that it's a dynamic quantize function. Or is there a deeper reason?

(my original comment on another PR was about reusing packing, less about naming as the naming of ops is internal)

**Comment by jianyuh** [comment]
> honestly, I'm not sure it's worth complexity and wrappers. We could just keep the original names, couldn't we? You can just add comment in native_function.yaml that it's a dynamic quantize function. Or is there a deeper reason?
> 
> (my original comment on another PR was about reusing packing, less about naming as the naming of ops is internal)

Sure. I can abandon this Diff / PR. This was originally based on the comments by @raghuramank100 on better distinguish the function names of static or dynamic quant in https://github.com/pytorch/pytorch/pull/22891, as the activation is fp32 for the dynamic quant FBGEMM wrapper here.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: Same here

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: If you are registering something with native_functions, you must be explicit which function you are calling here -- otherwise this will be ambiguous.

**Review by z-a-f — CHANGES_REQUESTED** [review]

**Comment by jamesr66a** [comment]
I will approve this if you address @jspark1105's comment about just calling the existing function

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: I mentioned it before -- single character variables are NOT ok! A general rule of thumb is, if it takes me more than a glance to understand what the variable is representing -- rename it. Every time I read this piece I have to go back and forth just to make sure there is no accidental typo bug.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: precompute `1 << (precision - 1)`?

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: Consistency: Either use `value /*variable*/` or `/*variable=*/ value`, but not both

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: Does increasing the number of threads provide a speedup?

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [inline aten/src/ATen/native/QuantizedLinear.cpp]: nit: `0.0f`

**Review by z-a-f — CHANGES_REQUESTED** [review]

**Comment by jianyuh** [comment]
> I will approve this if you address @jspark1105's comment about just calling the existing function

The issue is that when I call the existing function (e.g., call `fbgemm_linear_int8_weight_fp32_activation` inside `fbgemm_linear_int8_weight`), it shows the ambiguous function call errors:
```
aten/src/ATen/native/QuantizedLinear.cpp:173:10: error: call to 'fbgemm_linear_int8_weight_fp32_activation' is ambiguous
  return fbgemm_linear_int8_weight_fp32_activation(
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
aten/src/ATen/native/QuantizedLinear.cpp:405:10: error: call to 'fbgemm_linear_fp16_weight_fp32_activation' is ambiguous
  return fbgemm_linear_fp16_weight_fp32_activation(input, packed_weight, bias);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
stderr: caffe2/aten/src/ATen/native/QuantizedLinear.cpp:173:10: error: call to 'fbgemm_linear_int8_weight_fp32_activation' is ambiguous
  return fbgemm_linear_int8_weight_fp32_activation(
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
aten/src/ATen/native/QuantizedLinear.cpp:405:10: error: call to 'fbgemm_linear_fp16_weight_fp32_activation' is ambiguous
  return fbgemm_linear_fp16_weight_fp32_activation(input, packed_weight, bias);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I tried @zafartahirov 's suggestion that we should use anonymous namespace for these functions, but it didn't work.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: Sorry I might misunderstand it. How can I make the function call more explicit? I thought you want me to use the anonymous namespace to solve this issue, but it doesn't work.

**Comment by dzhulgakov** [comment]
Is it just namespace problem? maybe try `at::native::fbgemm_linear_fp16_weight_fp32_activation`?

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: I will follow your suggestion, but this should be able to be automatically precomputed by the compiler: precision is a `constexpr`, so `1 << (precision - 1)` is also a `constexpr`.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: Follow your suggestion.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: Fixed.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: For now, we focus on the single-thread performance with the quantization support. This is because normally we can run multiple instances on the servers. We can shift back to the intra-op parallelism after the PyTorch intra-op support is fully supported.

**Comment by jianyuh** [comment]
> Is it just namespace problem? maybe try `at::native::fbgemm_linear_fp16_weight_fp32_activation`?

Thanks for the insight! Right: this solution works!

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (jamesr66a) [reply aten/src/ATen/native/QuantizedLinear.cpp]: K and N are the conventional names for dimensions in all GEMM routines: https://spatial-lang.org/gemm

We should revert back to using those identifiers, otherwise this will cause confusion

Comment (jamesr66a) [reply aten/src/ATen/native/QuantizedLinear.cpp]: On second thought, this section is pretty insane:

```
      /*nRow=*/batch_size,
      /*nCol=*/input_channels,
      /*smat=*/input_ptr,
      /*ld=*/input_channels,
```

Can we instead rename these variables to match the FBGEMM API (i.e. `nRow`, `nCol`, and `ld)?

Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: `nRow`, `nCol` needs to be connected to a specific matrix. For `fbgemm::PackAWithQuantRowOffset` (`PackA`), `nRow` is `M` or `batch_size`, and `nCol` is `K` or `input_channels`; For `fbgemm::PackBMatrix`, `nRow` is `K` or `input_channels` and `nCol` is `N` or `output_channels`. The name `ld` is for the leading dimension (https://www.youtube.com/watch?v=PhjildK5oO8), which also depends on whether the matrix is A, B, or C, as well as the matrix is stored in the row-major or column-major. It might cause confusions if we rename these variables as `nRow`, `nCol`, and `ld`

Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: @jamesr66a : I totally concur with you that we can just use `M`, `K`, and `N`, as they are the conventional names for dimensions in all GEMM routines (https://petewarden.com/2015/04/20/why-gemm-is-at-the-heart-of-deep-learning/). @zafartahirov : What do you think?

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [reply aten/src/ATen/native/QuantizedLinear.cpp]: Sorry about the confusion. The problem with anonymous namespace is that it hides it from external use, but within the same file, it is visible. That makes it ambiguous if the function should be used from "nativeFunctions.h" or from the current file. If you explicitly specify the namespace (say `at::...`), it will be less confusing fro the compiler.

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (zafartahirov) [reply aten/src/ATen/native/QuantizedLinear.cpp]: The links to the blog posts are irrelevant in the case of the dev, because readability of the code is there to make the life of the devs easier, not the blog readers. The problem with names like `M`, `N`, `K` is that they force the readers to go back to the other parts of the code, just to clarify what was going on. The reason it is a standard within FBGEMM is not because it is good, but because it was not caught earlier in the development. 

However, it has a minimal impact on productivity. And although I am a strong proponent of descriptive variable names, I will submit to the better judgement of the FBGEMM devs (it doesn't mean I will not flag it in the future tho 😄 ).

**Review by z-a-f — APPROVED** [review]
LGTM

File: aten/src/ATen/native/QuantizedLinear.cpp
Comment (dzhulgakov) [reply aten/src/ATen/native/QuantizedLinear.cpp]: when we get to it - we can use `at::get_num_interop_threads` - but we'd need to share the thread pool with ATen's one

Comment (jianyuh) [reply aten/src/ATen/native/QuantizedLinear.cpp]: Will mark this as a TODO in the task. FBGEMM doesn't have the thread runtime: It only need the `thread_id` and `num_threads` for the task partitioning.

**Review by jamesr66a — APPROVED** [review]
I think this change is fine and makes it more explicit what the operator is actually doing. As discussed earlier, if you don't want to land it that's okay but it has my approval if you do

**Pull request closed without merging**
